src/java.base/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -398,20 +398,13 @@ private int checkArraySize(String typeName, int elemSize) {
     }
 
     @ForceInline
-    void checkBounds(long offset, long length) {
-        if (length > 0) {
-            Preconditions.checkIndex(offset, this.length - length + 1, this);
-        } else if (length < 0 || offset < 0 ||
-                offset > this.length - length) {
-            throw outOfBoundException(offset, length);
-        }
+    void checkSliceBounds(long offset, long length) {
+        Preconditions.checkFromIndexSize(offset, length, this.length, this::sliceOutOfBoundException);
```

Comment:
Is this performance-wise neutral? This is on a hot path and captures a lambda every time, need to verify compiler can eliminate this, which I doubt.